File: Client/Assets/Scripts/Utils/Timer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;

class Timer
{
	float length;
	public float Length { get { return length; } }

	float value;
	public float Value { get { return value; } }

	public bool Done { get { return value >= 1f; } }

	public Timer(float length, float value = 0f)
	{
		this.length = length;
		this.value = value;
	}

	public void Update(float delta)
	{
		value = Mathf.Clamp(value + delta / length, 0f, 1f);
	}

	public void Reset() { Reset(length); }
	public void Reset(float newLength)
	{
		length = newLength;
		value = 0;
	}
}

File: Server/Server/timer.cpp
#include "timer.hpp"

#include <glm/glm.hpp>

Timer::Timer( float length, float value ) :
	value( value ),
	length( length ) {
}

bool Timer::isDone( ) {
	return value >= 1.f;
}

void Timer::reset( ) { reset( length ); }
void Timer::reset( float length ) {
	this->length = length;
	value = 0.f;
}

void Timer::update( float delta ) {
	value = glm::clamp( value + delta / length, 0.f, 1.f );
}